admin/cli/cmd/base/options.go

```
@@ -0,0 +1,161 @@
+	if o.KeyVaultURI == "" {
+		return nil, fmt.Errorf("ga-auth-cert-kv cannot be empty")
+	}
+
+	if o.CertificateSecret == "" {
+		return nil, fmt.Errorf("ga-auth-cert-secret cannot be empty")
+	}
+
+	if o.TenantID == "" {
+		return nil, fmt.Errorf("ga-auth-tenant-id cannot be empty")
+	}
+
+	if o.ClientID == "" {
+		return nil, fmt.Errorf("ga-auth-client-id cannot be empty")
+	}
+
+	if len(o.Scopes) == 0 {
+		return nil, fmt.Errorf("ga-auth-scopes cannot be empty")
+	}
+
+	if o.Endpoint == "" {
+		return nil, fmt.Errorf("admin-api-endpoint cannot be empty")
+	}
+
+	return &ValidatedAuthTokenOptions{
+		validatedAuthOptions: &validatedAuthOptions{
+			RawAuthOptions: o,
+		},
+	}, nil
+}
+
+type validatedAuthOptions struct {
+	*RawAuthOptions
+}
+
+type ValidatedAuthTokenOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*validatedAuthOptions
+}
+
+type completedAuthOptions struct {
+	Token    string
+	Endpoint string
+	Host     string
+	Insecure bool
+}
+
+type AuthOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedAuthOptions
+}
+
+func (o *validatedAuthOptions) Complete(ctx context.Context) (*AuthOptions, error) {
+
+	// Login with the certificate
+	token, err := auth.GetBearerToken(ctx, auth.CertificateAuthConfig{
+		KeyVaultURL:           o.KeyVaultURI,
+		CertificateSecretName: o.CertificateSecret,
+		TenantID:              o.TenantID,
+		ClientID:              o.ClientID,
+		Scopes:                o.Scopes,
+		SendX5C:               true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get bearer token: %w", err)
+	}
+
+	return &AuthOptions{
+		completedAuthOptions: &completedAuthOptions{
+			Token:    token,
+			Endpoint: o.Endpoint,
+			Host:     o.Host,
+			Insecure: o.Insecure,
+		},
+	}, nil
+}
+
+func (o *AuthOptions) Execute(ctx context.Context) error {
+	logger := logr.FromContextOrDiscard(ctx)
```

prefer to pass logger explicitly, if you must not, at least handle the error. implicit parameters in context + panic is a good way to end up in a bad place

---

i handle the error. it is for sure better to fail than to silently drop logging and become blind when no logger was passed via the context